pkg/provisioning/import.go

```diff
@@ -0,0 +1,712 @@
+		return nil
+	}
+
+	// compare them again but ignore fields that can be updated, if configs
+	// are still different an update is not possible, we have to entirely
+	// recreate the connector
+	opts = []cmp.Option{
+		cmpopts.IgnoreFields(config.Connector{}, "Name", "Settings", "Processors"),
+	}
+	if cmp.Equal(oldConfig, newConfig, opts...) {
+		// only updatable fields don't match, we can update the connector
+		return []action{updateConnectorAction{
+			oldConfig:        oldConfig,
+			newConfig:        newConfig,
+			connectorService: s.connectorService,
+		}}
+	}
+
+	// we have to delete the old connector and create a new one
+	return []action{
+		deleteConnectorAction{
+			cfg:              oldConfig,
+			pipelineID:       pipelineID,
+			connectorService: s.connectorService,
+			pluginService:    s.pluginService,
+		},
+		createConnectorAction{
+			cfg:              newConfig,
+			pipelineID:       pipelineID,
+			connectorService: s.connectorService,
+			pluginService:    s.pluginService,
+		},
+	}
+}
+
+func (s *Service) prepareProcessorActions(oldConfig, newConfig config.Processor, parent processor.Parent) []action {
+	if oldConfig.ID == "" {
+		// no old config, it's a brand new processor
+		return []action{createProcessorAction{
+			cfg:              newConfig,
+			parent:           parent,
+			processorService: s.processorService,
+		}}
+	} else if newConfig.ID == "" {
+		// no new config, it's an old processor that needs to be deleted
+		return []action{deleteProcessorAction{
+			cfg:              oldConfig,
+			parent:           parent,
+			processorService: s.processorService,
+		}}
+	}
+
+	// first compare whole configs
+	if cmp.Equal(oldConfig, newConfig) {
+		// configs match, no need to do anything
+		return nil
+	}
+
+	// compare them again but ignore fields that can be updated, if configs
+	// are still different an update is not possible, we have to entirely
+	// recreate the processor
+	opts := []cmp.Option{
+		cmpopts.IgnoreFields(config.Processor{}, "Settings", "Workers"),
+	}
+	if cmp.Equal(oldConfig, newConfig, opts...) {
+		// only updatable fields don't match, we can update the processor
+		return []action{updateProcessorAction{
+			oldConfig:        oldConfig,
+			newConfig:        newConfig,
+			processorService: s.processorService,
+		}}
+	}
+
+	// we have to delete the old processor and create a new one
+	return []action{
+		deleteProcessorAction{
+			cfg:              oldConfig,
+			parent:           parent,
+			processorService: s.processorService,
+		},
+		createProcessorAction{
+			cfg:              newConfig,
+			parent:           parent,
+			processorService: s.processorService,
+		},
+	}
+}
+
+func (*Service) reverseActions(actions []action) {
+	for i, j := 0, len(actions)-1; i < j; i, j = i+1, j-1 {
+		actions[i], actions[j] = actions[j], actions[i]
+	}
+}
+
+type action interface {
+	String() string
+	Do(context.Context) error
+	Rollback(context.Context) error
+}
+
+// --------------------
```

Comment:
Nitpick: I think it would be good to have actions in a separate file.

Reply:
Sure, how does `import_actions.go` sound?

Reply:
Sounds 100% great to me!